### ISSUE TITLE
Lock order reversal

### ISSUE THREAD

**Issue by jasone** [jemalloc/jemalloc#719]
On a 14-core Ubuntu 16.04.2 system:
```
CC="gcc -m32" CXX="g++ -m32" ./autogen.sh --enable-debug --enable-prof && make -j32 all tests && make check
[...]
=== test/unit/prof_reset ===
test_prof_reset_basic: pass
test_prof_reset_cleanup: pass
<jemalloc>: Lock rank order reversal: prof_tdata(7) arenas(3)
Aborted (core dumped)
[...]
#0  0xf7761c89 in ?? ()
#1  0x0807ccca in jet_witness_lock (tsdn=0xf64f9a40, witness=0x80b5394 <arenas_lock+84>) at include/jemalloc/internal/witness_inlines.h:154
#2  0x0807d057 in jet_malloc_mutex_lock (tsdn=0xf64f9a40, mutex=0x80b5340 <arenas_lock>) at include/jemalloc/internal/mutex_inlines.h:65
#3  0x0804cea5 in jet_arena_init (tsdn=0xf64f9a40, ind=8, extent_hooks=0x809ef20 <jet_extent_hooks_default>) at src/jemalloc.c:417
#4  0x0804bdd7 in jet_arena_get (tsdn=0xf64f9a40, ind=8, init_if_missing=true) at include/jemalloc/internal/jemalloc_internal.h:916
#5  0x08052cd6 in jet_percpu_arena_update (tsd=0xf64f9a40, cpu=8) at include/jemalloc/internal/arena_inlines_a.h:55
#6  0x0804bf2d in jet_arena_choose_impl (tsd=0xf64f9a40, arena=0x0, internal=true) at include/jemalloc/internal/jemalloc_internal.h:976
#7  0x0804bf90 in jet_arena_ichoose (tsd=0xf64f9a40, arena=0x0) at include/jemalloc/internal/jemalloc_internal.h:992
#8  0x0805c423 in ckh_shrink (tsd=0xf64f9a40, ckh=0xf6cfdd2c) at src/ckh.c:316
#9  0x0805ca27 in jet_ckh_remove (tsd=0xf64f9a40, ckh=0xf6cfdd2c, searchkey=0xf6d4d7d4, key=0x0, data=0x0) at src/ckh.c:502
#10 0x080857cf in prof_tctx_destroy (tsd=0xf64f9a40, tctx=0xf3e7ca30) at src/prof.c:642
#11 0x08084e6b in jet_prof_free_sampled_object (tsd=0xf64f9a40, usize=8, tctx=0xf3e7ca30) at src/prof.c:250
#12 0x0807f6d2 in jet_prof_free (tsd=0xf64f9a40, ptr=0xf34ba300, usize=8) at include/jemalloc/internal/prof_inlines_b.h:226
#13 0x080507fc in ifree (tsd=0xf64f9a40, ptr=0xf34ba300, tcache=0xf6d39000, slow_path=true) at src/jemalloc.c:2010
#14 0x08051da4 in jet_dallocx (ptr=0xf34ba300, flags=0) at src/jemalloc.c:2618
#15 0x0804a7dc in thd_start (varg=0xffbbaa80) at test/unit/prof_reset.c:174
#16 0xf772926a in start_thread (arg=0xf64f9b40) at pthread_create.c:333
#17 0xf75e226e in clone () from /lib32/libc.so.6
```

**Comment by jasone** [comment]
Relevant rev: 5d33233a5e6601902df7cddd8cc8aa0b135c77b2

**Comment by jasone** [comment]
This was resolved by d16f1e53df3836f95deeca73419bb8c541aa579f.
